Allow removing implicit quadlet systemd dependencies

lambinoo opened the pull request:
Allow removing implicit quadlet systemd dependencies
    
Quadlet inserts network-online.target Wants/After dependencies to ensure pulling works.
Those systemd statements cannot be subsequently reset.

In the cases where those dependencies are not wanted, we add a new
configuration item called `DefaultDependencies=` in a new section called
[Quadlet]. This section is shared between different unit types.
    
fixes #24193

```release-note
Adds a directive in a new [Quadlet] section to prevent quadlet from adding implicit dependencies on the network-online target.
```

lambinoo commented:
> The code looks good. But, you will need to add some tests

Would an e2e test be enough for this?

mheon commented:
Sure. We generally want an E2E or system test, adding both is usually unnecessary.

lambinoo commented:
Should be good now I think

vrothberg reviewed (approved):
LGTM, nice work.

Luap99 commented on pkg/systemd/quadlet/quadlet.go:
```diff
@@ -509,10 +513,10 @@ func ConvertContainer(container *parser.UnitFile, isUser bool, unitsInfoMap map[
-	// override the default ones.
-	service.PrependUnitLine(UnitGroup, "After", "network-online.target")
-	service.PrependUnitLine(UnitGroup, "Wants", "network-online.target")
+	if service.LookupBooleanWithDefault(ContainerGroup, KeyQuadletDefaultDependencies, true) {
```
Do we really want to add these generic keys to each group? Overall this just adds a ton of duplication in the docs.
And every time we add such a generic option we have to do that over and over again which does not look sane to me.

So maybe is it time to define a `[Quadlet]` section and then just name the key `DefaultDependencies` under there? This would only need to be documented once and then should work for all units. @ygalblum WDYT?

vrothberg replied:
I had a similar thought but didn't consider the long-term impact on new features. +1 from my side.

ygalblum replied:
I agree and I was actually thinking about it for a while now. For example, I would move `ContainersConfModule` to there as well (I know, for backward compatibility we'll need to support both)
I'm just not sure we want to force @lambinoo to make this change.
Let's hold this PR for a sec and I'll try to find some time to add this section

lambinoo replied:
I don't mind. Just pushed the (what I think are) the required changes for this!

lambinoo commented:
Anyone has an idea why `hack/xref-quadlet-docs` thinks DefaultDependencies is out-of-order here when it's the only key in the table? I'm not really fluent in perl to be honest.

ygalblum commented on docs/source/markdown/podman-systemd.unit.5.md:
```diff
@@ -1914,6 +1922,23 @@ Override the default architecture variant of the container image.
+
+| **[Quadlet] options**      | **Description**                                   |
+|----------------------------|---------------------------------------------------|
+| DefaultDependencies=false  | Include implicit network dependencies to the unit |
```
The example sets the value to `false` but the description refers to the case of `true`. This might be a bit confusing

ygalblum replied:
Something got reverted here and the original issue is back. I'll take advantage and say that since setting it to `true` is redundant, it might be better to leave the `false` value and change the description to something like "Disable implicit network..."

ygalblum commented on pkg/systemd/quadlet/quadlet.go:
```diff
@@ -38,6 +38,7 @@ const (
 	VolumeGroup     = "Volume"
 	ImageGroup      = "Image"
 	BuildGroup      = "Build"
+	QuadletGroup    = "Quadlet"
```
You need to also define `XQuadletGroup    = "X-Quadlet"` and make sure to rename the group same as in https://github.com/containers/podman/blob/90fd1dce3420133bd6e472872ccfaa49dfe53f19/pkg/systemd/quadlet/quadlet.go#L542

lambinoo replied:
Done in every unit types

ygalblum reviewed (commented):
Additional question:
The new `Quadlet` section is currently only relevant for `.container`, `.build` and `.image` files. What should be the behaviour if it appears on a different types? I see three options:

1. Rename the group to `X-Quadlet` and noop
2. Error at the Quadlet level
3. Noop which will result in a systemd error because of an unrecognized section

I think option 1 is preferable as it will also be future proof.

lambinoo commented:
> Additional question: The new `Quadlet` section is currently only relevant for `.container`, `.build` and `.image` files. What should be the behaviour if it appears on a different types? I see three options:
> 
>     1. Rename the group to `X-Quadlet` and noop
> 
>     2. Error at the Quadlet level
> 
>     3. Noop which will result in a systemd error because of an unrecognized section
> 
> 
> I think option 1 is preferable as it will also be future proof.

+1 on renaming. The quadlet section will be used by `.pod` too for example when `ContainersConfModule=` is added to it, as you mentioned before.

ygalblum reviewed (approved):
LGTM, Thanks!

lambinoo commented:
Is there a way to retry those podman db test failures? They seem pretty unrelated to quadlet

Luap99 commented:
> Is there a way to retry those podman db test failures? They seem pretty unrelated to quadlet. And I'm not too familiar with Github actions

We can restart individual ones yes, but if basically all of them are red the problem is in the new changes (expect in very rare cases where some external factor is breaking things)
You can click on details and then open the link at the bottom which should show you the error
https://api.cirrus-ci.com/v1/artifact/task/4942923031117824/html/int-podman-debian-13-root-host-sqlite.log.html#t--quadlet-system-generator-Running-quadlet-dryrun-tests-Should-fail-on-bad-quadlet--1

So there is definitely something broken in quadlet that makes this this fail

lambinoo commented:
> > Is there a way to retry those podman db test failures? They seem pretty unrelated to quadlet. And I'm not too familiar with Github actions
> 
> We can restart individual ones yes, but if basically all of them are red the problem is in the new changes (expect in very rare cases where some external factor is breaking things) You can click on details and then open the link at the bottom which should show you the error https://api.cirrus-ci.com/v1/artifact/task/4942923031117824/html/int-podman-debian-13-root-host-sqlite.log.html#t--quadlet-system-generator-Running-quadlet-dryrun-tests-Should-fail-on-bad-quadlet--1
> 
> So there is definitely something broken in quadlet that makes this this fail

Oh, good catch yeah. I did see a lot of those weird "unlikeat" errors before and stopped looking after seeing one. I'll take a look at them this evening, or tomorrow if I don't have time

lambinoo commented:
Looks a bit better, but I can see a pasta related error and a WSL one.

ygalblum commented on docs/source/markdown/podman-systemd.unit.5.md:
```diff
@@ -1914,6 +1922,23 @@ Override the default architecture variant of the container image.
+
+Add Quadlet's default dependencies to the unit.
+
+If set to true (the default), Quadlet will add a dependency (After=, Wants=) to `network-online.target` to the generated unit.
```
Also here. Since `true` is the default, I think it's better to describe the outcome of setting it to `false`

lambinoo replied:
Updated for both!

ygalblum reviewed (changes requested):
Your latest forced push reverted all the changes made following the comments. I added a comment about the doc

lambinoo commented:
> Your latest forced push reverted all the changes made following the comments. I added a comment about the doc

Added back everything. The cons on working on two different workstations and badly keeping track..

ygalblum reviewed (approved):
LGTM

Luap99 reviewed (approved):
/lgtm
thanks
